py_ocpi/credentials/v_2_2_1/api/cpo.py

```diff
@@ -1,10 +1,113 @@
-from fastapi import APIRouter
+import uuid
+import httpx
+
+from fastapi import APIRouter, Depends, HTTPException, Request, status as fastapistatus
+from pydantic import ValidationError
+
+from py_ocpi.core.schemas import OCPIResponse
+from py_ocpi.crud import get_crud
+from py_ocpi.adapter import get_adapter
+from py_ocpi.core import status
+from py_ocpi.core.enums import ModuleID
+from py_ocpi.credentials.v_2_2_1.schemas import Credentials, ServerCredentials
 
 router = APIRouter(
     prefix='/credentials',
 )
 
 
-@router.get("/")
-def get_credentials():
-    return "credentials app created!"
+@router.get("/", response_model=OCPIResponse)
+async def get_credentials(request: Request, crud=Depends(get_crud), adapter=Depends(get_adapter)):
+    headers = request.headers
+    headers_token = headers.get('authorization')
+    token = headers_token.split()[1]
+    try:
+        data = await crud.get(ModuleID.CredentialsAndRegistrations, token)
+        return OCPIResponse(
+            data=[adapter.credentials_adapter(data).dict()],
+            **status.OCPI_1000_GENERIC_SUCESS_CODE,
+        )
+    except ValidationError:
+        return OCPIResponse(
+            data=[],
+            **status.OCPI_3001_UNABLE_TO_USE_CLIENTS_API,
+        )
+
+
+@router.post("/", response_model=OCPIResponse)
+async def post_credentials(credentials: Credentials, crud=Depends(get_crud), adapter=Depends(get_adapter)):
+    try:
+        # Check if the client is already registered
+        credentials_client_token = credentials.token
+        server_cred = await crud.get(ModuleID.CredentialsAndRegistrations, credentials_client_token)
+        if server_cred:
+            raise HTTPException(fastapistatus.HTTP_405_METHOD_NOT_ALLOWED, "Client is already registered")
+
+        # Retrieve the versions and endpoints from the client
+        async with httpx.AsyncClient() as client:
+            authorization_token = f'Token {credentials_client_token}'
+            response_versions = await client.get(credentials.url,
+                                                 headers={'authorization': authorization_token})
+            versions = response_versions.json()['data'][0]
+            response_endpoints = await client.get(versions['url'],
+                                                  headers={'authorization': authorization_token})
+
+        if response_endpoints.status_code == fastapistatus.HTTP_200_OK:
+            # Store client credentials
+            endpoints = response_endpoints.json()['data'][0]
+            await crud.create(ModuleID.CredentialsAndRegistrations, ServerCredentials(
+                cred_token_b=credentials.token,
+                versions=versions,
+                endpoints=endpoints
+            ))
+
+            # Generate new credentials for sender
+            new_credentials = await crud.create(ModuleID.CredentialsAndRegistrations, Credentials(url=versions['url']))
```

Comment:
you need to pass a dict instead of `Credentials(url=versions['url'])`. because Credentials will expect token and roles parameters